src/formik.tsx

```diff
@@ -248,42 +254,40 @@ export function Formik<Props, Values extends FormikValues, Payload>({
         this.setState({ isSubmitting });
       };
 
-      runValidationSchema = (values: Values, cb?: Function) => {
-        validateFormData<Values>(values, validationSchema).then(
+      /**
+       * Run validation against a Yup schema and optionally run a function if successful
+       */
+      runValidationSchema = (values: Values, onSuccess?: Function) => {
+        validateYupSchema<Values>(values, validationSchema).then(
           () => {
             this.setState({ errors: {} });
-            if (cb) {
-              cb();
+            if (onSuccess) {
+              onSuccess();
             }
           },
           (err: any) =>
             this.setState({ errors: yupToFormErrors(err), isSubmitting: false })
         );
       };
 
-      runValidationAsync = (values: Values, cb?: Function) => {
-        validateAsync(values, this.props).then(
-          () => {
-            this.setState({ errors: {} });
-            if (cb) {
-              cb();
-            }
-          },
-          errors => this.setState({ errors, isSubmitting: false })
-        );
-      };
-
-      runValidationSync = (values: Values) => {
-        this.setErrors(validate(values));
-      };
-
+      /**
+       * Run validations and update state accordingly
+       */
       runValidations = (values: Values) => {
         if (validate) {
-          this.runValidationSync(values);
-        }
-        if (validateAsync) {
-          this.runValidationAsync(values);
+          const maybePromisedErrors = validate(values, this.props);
+          if (isPromise(maybePromisedErrors)) {
+            (validate(values, this.props) as Promise<any>).then(
```

Comment:
Second call to `(validate(values, this.props) as Promise<any>)` causes second validation run here. It feasible to run `maybePromisedErrors.then()` here - not sure whether you loose some of typehints here...